### PR TITLE
Add `pytorch-dev-deps`

### PR DESCRIPTION
**Pull request by pmeier**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/pytorch-dev-deps/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package:
+  name: pytorch-dev-deps
+  version: 1.0
+
+build:
+  noarch: python
+  number: 0
+  script:
+    - install -Dm664 {{ RECIPE_DIR }}/activate.sh {{ PREFIX }}/etc/conda/activate.d/activate-pytorch-dev-deps.sh  # [unix]
```
Comment (pmeier) [inline recipes/pytorch-dev-deps/meta.yaml:9]: We also need an activation script for Windows.

File: recipes/pytorch-dev-deps/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+test:
+  commands:
+    - echo "Environment was successfully solved!"
+
+about:
+  summary: Development dependencies to work on PyTorch.
+  home: https://github.com/pytorch/pytorch
+  license: BSD-3-Clause
+  license_family: BSD
+  license_file: LICENSE
```
Comment (pmeier) [inline recipes/pytorch-dev-deps/meta.yaml:39]: What license do we use? Currently the file is just a dummy to make CI happy.

File: recipes/pytorch-dev-deps/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    - mkl-include
+    - setuptools
+    - cmake
+    - cffi
+    - typing
+    - pytest
+    - compilers
+    - flake8
+    - psutil
+    - hypothesis
```
Comment (pmeier) [inline recipes/pytorch-dev-deps/meta.yaml:27]: Do we want to pin these to specific versions in any way?

**Review by pmeier — COMMENTED** [review]
@pearu Please review. I was able to build with only this installed. Feel free to make changes. The branch is part of the [`Quansight/conda-forge-staged-recipes` repo](https://github.com/Quansight/conda-forge-staged-recipes/tree/pytorch-dev-deps).

Besides my inline comments, I have one further question:

Do we want to provide `install_cuda_deps.sh` scripts that detect the available CUDA version and installs the corresponding magma and nvcc bindings?

File: recipes/pytorch-dev-deps/meta.yaml (hunk 1 above)
Comment (pmeier) [reply recipes/pytorch-dev-deps/meta.yaml:9]: Apparently, macOS also needs different commands.

**Comment by pmeier** [comment]
@pearu 

Not sure what to do about

> Non noarch packages should have python requirement without any version constraints.

If I understand the [requirements for `noarch` option](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python) correctly, we cannot use it in case we have activation scripts. Thus, we probably need to remove the `>=3.6` constraint from the `python` dependency. If I'm not mistaken that would make `conda install -c conda-forge python==2.7 pytorch-dev-deps` valid although this will not work out.

**Comment by pmeier** [comment]
It seems PyTorch's test suite has a hard dependency on [`expecttest`](https://pypi.org/project/expecttest/). It is not available on `conda-forge` so I need to package it first.

**Comment by pmeier** [comment]
> It seems PyTorch's test suite has a hard dependency on [`expecttest`](https://pypi.org/project/expecttest/). It is not available on `conda-forge` so I need to package it first.

Done: https://github.com/conda-forge/expecttest-feedstock

File: recipes/pytorch-dev-deps/meta.yaml (hunk 3 above)
Comment (pearu) [reply recipes/pytorch-dev-deps/meta.yaml:24]: No, unless there is a specific need for a pin: we want to minimize the maintenance burden that pinning would introduce.

File: recipes/pytorch-dev-deps/meta.yaml (hunk 2 above)
Comment (pearu) [reply recipes/pytorch-dev-deps/meta.yaml:37]: BSD-3 sounds good to me.

File: recipes/pytorch-dev-deps/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package:
+  name: pytorch-dev-deps
+  version: 1.0
```
[the remark this answers is not in the document]
Comment (pearu) [reply recipes/pytorch-dev-deps/meta.yaml:3]: `pytorch-dev-deps` ought to be compatible with the PyTorch master branch.
So, how about using the following policy:
- any maintenance change to `pytorch-dev-deps` (e.g. bug fix) that is independent of PyTorch master branch state, increase the `build` number.
- when some merged PyTorch PR requires a change to `pytorch-dev-deps`, increase the major `version` number (and reset `build` to `0` and minor `version` to `0`).
- when `pytorch-dev-deps` introduces a new feature (independent of PyTorch master branch state), increase the minor `version` number (and reset `build` to `0`).

[the remark this answers is not in the document]
Comment (pmeier) [reply recipes/pytorch-dev-deps/meta.yaml:3]: This sounds good in general. Do you an example for the minor version change though? What features do you think we can introduce here that are not bug fixes are dependency changes?

[the remark this answers is not in the document]
Comment (pearu) [reply recipes/pytorch-dev-deps/meta.yaml:3]: An example would be when pytorch-dev-deps introduces new outputs. The outputs can be used (but not necessarily) to support cpu-only and cuda-enabled environments, for instance.

**Comment by pearu** [comment]
I accept to be a maintainer of this conda-forge package.

**Comment by pearu** [comment]
> > Non noarch packages should have python requirement without any version constraints.
> 
> If I understand the [requirements for `noarch` option](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python) correctly, we cannot use it in case we have activation scripts. Thus, we probably need to remove the `>=3.6` constraint from the `python` dependency. If I'm not mistaken that would make `conda install -c conda-forge python==2.7 pytorch-dev-deps` valid although this will not work out.

IIUC, this package will not be a noarch package. For instance, we want to introduce activate scripts.

**Comment by pmeier** [comment]
> IIUC, this package will not be a noarch package. For instance, we want to introduce activate scripts.

Yes, it will be a `noarch` package. But for these we can't pin `python` to `>= 3.6` as we do now. I had a chat on Gitter with @wolfv about this and it is safe to remove our pin as the build process will automatically inject it back in.

**Pull request closed without merging**